### PR TITLE
Add static configuration for Namespaces [HZ-2576]

### PR DESCRIPTION
**Pull request by JamesHazelcast**
This PR focuses on the changes required for the static configuration of Namespaces.
[Namespaces feature TDD for reference](https://gist.github.com/JamesHazelcast/f25007a4e0594c491ce867c816b616a7).

_This PR is 1 of 5 pertaining to the Namespace feature addition. The other 4 PRs can be found below:_
- https://github.com/hazelcast/hazelcast/pull/26145
- https://github.com/hazelcast/hazelcast/pull/26146
- https://github.com/hazelcast/hazelcast/pull/26147
- https://github.com/hazelcast/hazelcast/pull/26148

_Due to the Namespaces feature being split into 5 pieces for easier reviewing, it is highly recommended that you clone the [UCD feature development branch](https://github.com/vbekiaris/hazelcast/tree/enhancements/5.4.0/ucd) for testing purposes, as these individual PRs are likely non-functional standalone._

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/config/YamlMemberDomConfigProcessor.java [hunk 1]
```diff
@@ -899,6 +946,18 @@ void handleFencedLocks(CPSubsystemConfig cpSubsystemConfig, Node node) {
...
+//        for (Node child : childElements(node)) {
+//            Node named = getNamedItemNode(child, "name");
+//            NamespaceConfig nsConfig = new NamespaceConfig(named.getNodeValue());
+//            Node resources = getNamedItemNode(child, "resources");
+//            for (Node subC : childElements(resources)) {
+//                handleResources(subC, nsConfig);
+//            }
+//        }
+//    }
+
```
Comment (kwart) [inline hazelcast/src/main/java/com/hazelcast/internal/config/YamlMemberDomConfigProcessor.java:960]: a leftover?

**Review by kwart — COMMENTED** [review]
LGTM permission-wise.

**Review by vbekiaris — COMMENTED** [review]
I added a few comments. Feel free to address naming changes in a separate PR (1) after consensus on feature name with documentation team & (2) to avoid wreaking havoc on the chain of PRs.

File: hazelcast/src/main/java/com/hazelcast/internal/config/ScheduledExecutorConfigReadOnly.java [hunk 2]
```diff
@@ -59,4 +61,9 @@ public ScheduledExecutorConfig setMergePolicyConfig(MergePolicyConfig mergePolic
     public ScheduledExecutorConfig setStatisticsEnabled(boolean statisticsEnabled) {
         throw new UnsupportedOperationException("This config is read-only scheduled executor: " + getName());
     }
+
+    @Override
+    public ScheduledExecutorConfig setNamespace(@Nullable String namespace) {
+        throw new UnsupportedOperationException("This config is read-only scheduled executor: " + getName());
```
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/internal/config/ScheduledExecutorConfigReadOnly.java:67]: nit: These could be unified across configs - sometimes there is `throwReadOnly`, sometimes it is like this.

File: hazelcast/src/test/java/com/hazelcast/config/NamespaceConfigTest.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+import static com.hazelcast.test.HazelcastTestSupport.assumeDifferentHashCodes;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class NamespaceConfigTest {
+    @Test
+    public void testEqualsAndHashCode() {
+        assumeDifferentHashCodes();
+        EqualsVerifier.forClass(NativeMemoryConfig.class).suppress(Warning.NULL_FIELDS, Warning.NONFINAL_FIELDS).verify();
```
Comment (frant-hartm) [inline hazelcast/src/test/java/com/hazelcast/config/NamespaceConfigTest.java:36]: This seems to be a typo?

File: hazelcast/src/test/resources/hazelcast-fullconfig-advanced-network-config.xml [hunk 4]
```diff
@@ -36,7 +36,7 @@
 <hazelcast xmlns="http://www.hazelcast.com/schema/config"
            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://www.hazelcast.com/schema/config
-           http://www.hazelcast.com/schema/config/hazelcast-config-5.3.xsd">
+           http://www.hazelcast.com/schema/config/hazelcast-config-5.4.xsd">
```
Comment (frant-hartm) [inline hazelcast/src/test/resources/hazelcast-fullconfig-advanced-network-config.xml:39]: It seems we didn't do the usual update to next minor version, @vbekiaris should we do that in separate PR for all configs?

**Review by frant-hartm — CHANGES_REQUESTED** [review]
is it possible to replace statically defined configs?

**Comment by JamesHazelcast** [comment]
> is it possible to replace statically defined configs?

Yes, it should be - there was an oversight that prevented this, caught by Vassilis: https://github.com/hazelcast/hazelcast/pull/26145#discussion_r1424947649

File: hazelcast/src/main/java/com/hazelcast/config/NamespaceConfig.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+    protected NamespaceConfig add(ResourceDefinition resourceDefinition) {
+        if (resourceDefinitions.putIfAbsent(resourceDefinition.id(), resourceDefinition) != null) {
+            throw new IllegalArgumentException("Resource with id: " + resourceDefinition.id() + " already exists");
+        } else {
+            return this;
+        }
+    }
+
+    public Set<ResourceDefinition> getResourceConfigs() {
```
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/config/NamespaceConfig.java:100]: Addressed in: https://github.com/vbekiaris/hazelcast/commit/3ba9010e5b891308dd4c808cac78075d54391a8c

File: hazelcast/src/main/java/com/hazelcast/config/NamespacesConfig.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+package com.hazelcast.config;
+
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+/** @since 5.4 */
```
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/config/NamespacesConfig.java:25]: Addressed in: https://github.com/vbekiaris/hazelcast/commit/3ba9010e5b891308dd4c808cac78075d54391a8c

File: hazelcast/src/main/java/com/hazelcast/config/NamespaceConfig.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class NamespaceConfig implements NamedConfig, IdentifiedDataSerializable {
```
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/config/NamespaceConfig.java:39]: Addressed in: https://github.com/vbekiaris/hazelcast/commit/3ba9010e5b891308dd4c808cac78075d54391a8c

I've been using `@since 5.4` and looking around the codebase there's a mix of `X.Y` and `X.Y.Z` declarations - any preference here? Worth updating to match?

File: hazelcast/src/main/java/com/hazelcast/config/Config.java [hunk 8]
```diff
@@ -3229,6 +3231,24 @@ public DataConnectionConfig findDataConnectionConfig(String name) {
...
+    /**
+     * @return the namespaces configuration object
+     * @since 5.4.0
+     */
+    public NamespacesConfig getNamespacesConfig() {
+        return namespacesConfig;
+    }
+
+    /**
+     * Sets the namespaces configuration .
```
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/config/Config.java:3243]: Addressed in: https://github.com/vbekiaris/hazelcast/commit/3ba9010e5b891308dd4c808cac78075d54391a8c

File: hazelcast/src/main/java/com/hazelcast/config/Config.java [hunk 9]
```diff
@@ -3229,6 +3231,24 @@ public DataConnectionConfig findDataConnectionConfig(String name) {
...
+     *
+     * @since 5.4.0
+     */
+    public @Nonnull Config setNamespacesConfig(@Nonnull NamespacesConfig namespacesConfig) {
```
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/config/Config.java:3247]: Addressed in: https://github.com/vbekiaris/hazelcast/commit/3ba9010e5b891308dd4c808cac78075d54391a8c

File: hazelcast/src/main/java/com/hazelcast/config/NamespaceConfig.java [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    @Nullable
+    private String name;
```
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/config/NamespaceConfig.java:41]: Agreed it should not be nullable - I've also added a null check to the `setName()` method. Addressed in: https://github.com/vbekiaris/hazelcast/commit/3ba9010e5b891308dd4c808cac78075d54391a8c

File: hazelcast/src/main/java/com/hazelcast/config/NamespacesConfig.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+public class NamespacesConfig {
+    private boolean enabled;
+    private final Map<String, NamespaceConfig> namespaceConfigs = new ConcurrentHashMap<>();
+    private @Nullable JavaSerializationFilterConfig javaSerializationFilterConfig;
```
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/config/NamespacesConfig.java:29]: Agreed, although painful, I updated all references to be `class-filter` or `classFilterConfig` instead :) Addressed in: https://github.com/vbekiaris/hazelcast/commit/3ba9010e5b891308dd4c808cac78075d54391a8c

File: hazelcast/src/test/java/com/hazelcast/config/YamlConfigBuilderTest.java [hunk 12]
```diff
@@ -49,9 +51,12 @@
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
+import wiremock.com.google.common.collect.Sets;
```
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/test/java/com/hazelcast/config/YamlConfigBuilderTest.java:54]: Addressed in: https://github.com/vbekiaris/hazelcast/commit/3ba9010e5b891308dd4c808cac78075d54391a8c

File: hazelcast/src/main/java/com/hazelcast/config/NamespaceAwareConfig.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.config;
+
+import javax.annotation.Nullable;
+
+public interface NamespaceAwareConfig {
```
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/config/NamespaceAwareConfig.java:21]: Addressed in: https://github.com/vbekiaris/hazelcast/commit/3ba9010e5b891308dd4c808cac78075d54391a8c

File: hazelcast/src/main/java/com/hazelcast/config/NamespaceConfig.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/config/NamespaceConfig.java:39]: Addressed in: https://github.com/vbekiaris/hazelcast/commit/3ba9010e5b891308dd4c808cac78075d54391a8c Not sure why checkstyle didn't raise this, we ran the entire codebase through verification before these PRs were created.

File: hazelcast/src/main/java/com/hazelcast/config/NamespacesConfig.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+package com.hazelcast.config;
+
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+/** @since 5.4 */
+public class NamespacesConfig {
```
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/config/NamespacesConfig.java:26]: Addressed in: https://github.com/vbekiaris/hazelcast/commit/3ba9010e5b891308dd4c808cac78075d54391a8c

File: hazelcast/src/main/java/com/hazelcast/config/NamespaceAwareConfig.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+     * during operations related to the structure associated with this configuration.
+     *
+     * @return Namespace Name for use with the {@link com.hazelcast.internal.namespace.NamespaceService},
+     *         or {@code null} if there is no Namespace to associate with.
+     * @since 5.4
+     */
+    @Nullable
+    default String getNamespace() {
+        return DEFAULT_NAMESPACE;
+    }
```
[the remark this answers is not in the document]
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/config/NamespaceAwareConfig.java:47]: Addressed in: https://github.com/vbekiaris/hazelcast/commit/3ba9010e5b891308dd4c808cac78075d54391a8c

File: hazelcast/src/main/java/com/hazelcast/internal/config/ScheduledExecutorConfigReadOnly.java (hunk 2 above)
Comment (JamesHazelcast) [reply hazelcast/src/main/java/com/hazelcast/internal/config/ScheduledExecutorConfigReadOnly.java:67]: Agreed, but I'll leave that out of the UCD PR for now - I did think the same when I was looking over the PR changes.

File: hazelcast/src/test/java/com/hazelcast/config/NamespaceConfigTest.java (hunk 3 above)
Comment (JamesHazelcast) [reply hazelcast/src/test/java/com/hazelcast/config/NamespaceConfigTest.java:36]: Yes, looks like this test wasn't fully implemented. I've updated it to actually test `NamespaceConfig` functionality in: https://github.com/vbekiaris/hazelcast/commit/3ba9010e5b891308dd4c808cac78075d54391a8c

**Review by vbekiaris — APPROVED** [review]

**Review by frant-hartm — APPROVED** [review]

**Comment by JamesHazelcast** [comment]
Closing this PR due to the creation of the final mono-repo PR: https://github.com/hazelcast/hazelcast-mono/pull/217 🎉
